middle_end/flambda/types/env/typing_env_level.rec.ml

```diff
@@ -259,20 +276,31 @@ let join_types ~env_at_fork envs_with_levels =
             | None, Some use_ty ->
               (* In this case, we haven't yet got a joined type for [name]. *)
               let left_ty =
-                (* If this is the first level in the join, we just need to
-                   make the type suitable for the joined environment, so we
-                   use [Bottom] to avoid losing precision... *)
-                if is_first_join then Type_grammar.bottom_like use_ty
-                (* ...but if this is not the first level in the join, then we
+                (* If this is the first occurrence of [name] in the join,
+                   and [name] was not part of the base environment,
+                   we just need to make the type suitable for the joined
+                   environment, so we use [Bottom] to avoid losing precision...
+                *)
+                let is_first_definition =
+                  match Name.must_be_var_opt name with
+                  | None -> false
```

Comment:
assert false

Reply:
fatal errors are better: in particular, if I recall correctly, an `Assertion_failure _`, as raised by `assert false` is *not* caught by the handlers that print the flambda context, ^^